File: Android/tuilivekit/livekit/src/main/java/com/trtc/uikit/livekit/component/audioeffect/AudioEffectPanel.java
package com.trtc.uikit.livekit.component.audioeffect;


import android.annotation.SuppressLint;
import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.FrameLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.appcompat.widget.SwitchCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.trtc.uikit.livekit.R;
import com.trtc.uikit.livekit.common.DataReporter;
import com.trtc.uikit.livekit.component.audioeffect.service.AudioEffectService;
import com.trtc.uikit.livekit.component.audioeffect.store.AudioEffectState;
import com.trtc.uikit.livekit.component.audioeffect.store.AudioEffectStore;
import com.trtc.uikit.livekit.component.audioeffect.view.ChangeVoiceAdapter;
import com.trtc.uikit.livekit.component.audioeffect.view.ReverbAdapter;

@SuppressLint("ViewConstructor")
public class AudioEffectPanel extends FrameLayout {

    private static final int LIVEKIT_METRICS_PANEL_SHOW_LIVE_ROOM_AUDIO_EFFECT  = 190017;
    private static final int LIVEKIT_METRICS_PANEL_SHOW_VOICE_ROOM_AUDIO_EFFECT = 191015;

    private final Context            mContext;
    private       TextView           mTextEarReturnVolume;
    private       TextView           mTextPeopleVolume;
    private       AudioEffectService mAudioEffectService;
    private       AudioEffectState   mAudioEffectState;

    protected OnBackButtonClickListener mOnBackButtonClickListener;

    public AudioEffectPanel(Context context) {
        this(context, null);
    }

    public AudioEffectPanel(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public AudioEffectPanel(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mContext = context;
        LayoutInflater.from(mContext).inflate(R.layout.audio_effect_layout_settings_panel, this, true);
    }

    public void init(String roomId) {
        AudioEffectStore.sharedInstance().init(roomId);
        mAudioEffectState = AudioEffectStore.sharedInstance().mAudioEffectState;
        mAudioEffectService = AudioEffectStore.sharedInstance().mAudioEffectService;
        initView();
        reportData(roomId);
    }

    protected void initView() {
        bindViewId();

        initBackView();
        initChangeVoiceView();
        initReverbView();
        initEarReturnVolumeView();
        initPeopleVolumeView();
        initEnableEarReturnView();
        initEarReturnVolume();
    }

    private void bindViewId() {
        mTextEarReturnVolume = findViewById(R.id.tv_ear_return_volume);
        mTextPeopleVolume = findViewById(R.id.tv_people_volume);
    }

    private void initEarReturnVolume() {
        SeekBar seekEarReturnVolume = findViewById(R.id.sb_ear_return_volume);
        seekEarReturnVolume.setProgress(mAudioEffectState.earMonitorVolume.getValue());

        seekEarReturnVolume.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                mTextEarReturnVolume.setText(String.valueOf(i));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                mAudioEffectService.setVoiceEarMonitorVolume(seekBar.getProgress());
            }
        });
    }

    private void initEnableEarReturnView() {
        SwitchCompat switchEarReturn = findViewById(R.id.sc_enable_ear);
        switchEarReturn.setChecked(mAudioEffectState.enableVoiceEarMonitor.getValue());
        switchEarReturn.setOnCheckedChangeListener(
                (button, enable) -> mAudioEffectService.enableVoiceEarMonitor(enable));
    }

    private void initPeopleVolumeView() {
        mTextPeopleVolume.setText(String.valueOf(mAudioEffectState.voiceVolume.getValue()));

        SeekBar seekPeopleVolume = findViewById(R.id.sb_people_volume);
        seekPeopleVolume.setProgress(mAudioEffectState.voiceVolume.getValue());
        seekPeopleVolume.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                mTextPeopleVolume.setText(String.valueOf(i));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                mAudioEffectService.setVoiceVolume(seekBar.getProgress());
            }
        });
    }

    private void initEarReturnVolumeView() {
        mTextEarReturnVolume.setText((String.valueOf(mAudioEffectState.earMonitorVolume.getValue())));
    }

    private void initBackView() {
        findViewById(R.id.iv_back).setOnClickListener((view) -> {
            if (mOnBackButtonClickListener != null) {
                mOnBackButtonClickListener.onClick();
            }
        });
    }

    private void initReverbView() {
        RecyclerView recyclerReverb = findViewById(R.id.rv_reverb);
        ReverbAdapter adapterReverb = new ReverbAdapter(mContext);
        int spanCount = adapterReverb.getItemCount();
        recyclerReverb.setLayoutManager(new GridLayoutManager(mContext, spanCount));
        recyclerReverb.setAdapter(adapterReverb);
    }

    private void initChangeVoiceView() {
        RecyclerView recyclerChangeVoice = findViewById(R.id.rv_change_voice);
        ChangeVoiceAdapter adapterChangeVoice = new ChangeVoiceAdapter(mContext);
        int spanCount = adapterChangeVoice.getItemCount();
        recyclerChangeVoice.setLayoutManager(new GridLayoutManager(mContext, spanCount));
        recyclerChangeVoice.setAdapter(adapterChangeVoice);
    }

    public void setOnBackButtonClickListener(OnBackButtonClickListener listener) {
        mOnBackButtonClickListener = listener;
    }

    public interface OnBackButtonClickListener {
        void onClick();
    }

    private void reportData(String roomId) {
        boolean isVoiceRoom = !TextUtils.isEmpty(roomId) && roomId.startsWith("voice_");
        if (isVoiceRoom) {
            DataReporter.reportEventData(LIVEKIT_METRICS_PANEL_SHOW_VOICE_ROOM_AUDIO_EFFECT);
        } else {
            DataReporter.reportEventData(LIVEKIT_METRICS_PANEL_SHOW_LIVE_ROOM_AUDIO_EFFECT);
        }
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/audioeffect/AudioEffectPanel.kt
package com.trtc.uikit.livekit.component.audioeffect

import android.annotation.SuppressLint
import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.SeekBar
import android.widget.TextView
import androidx.appcompat.widget.SwitchCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.trtc.uikit.livekit.R
import com.trtc.uikit.livekit.common.reportEventData
import com.trtc.uikit.livekit.common.ui.BasicView
import com.trtc.uikit.livekit.component.audioeffect.view.ChangeVoiceAdapter
import com.trtc.uikit.livekit.component.audioeffect.view.ReverbAdapter
import com.trtc.uikit.livekit.component.audioeffect.viewmodel.AudioEffectConstants.LIVEKIT_METRICS_PANEL_SHOW_LIVE_ROOM_AUDIO_EFFECT
import com.trtc.uikit.livekit.component.audioeffect.viewmodel.AudioEffectConstants.LIVEKIT_METRICS_PANEL_SHOW_VOICE_ROOM_AUDIO_EFFECT
import io.trtc.tuikit.atomicxcore.api.AudioEffectStore
import io.trtc.tuikit.atomicxcore.api.DeviceStore

@SuppressLint("ViewConstructor")
class AudioEffectPanel @JvmOverloads constructor(
    private val context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : BasicView(context, attrs, defStyleAttr) {

    private lateinit var textEarReturnVolume: TextView
    private lateinit var textPeopleVolume: TextView
    private var audioEffectStore: AudioEffectStore? = null
    private var deviceStore: DeviceStore? = null
    private var onBackButtonClickListener: OnBackButtonClickListener? = null

    init {
        LayoutInflater.from(context)
            .inflate(R.layout.audio_effect_layout_settings_panel, this, true)
    }

    override fun init(roomId: String) {
        super.init(roomId)
        initView()
        reportData(roomId)
    }

    fun setOnBackButtonClickListener(listener: OnBackButtonClickListener?) {
        onBackButtonClickListener = listener
    }

    override fun initStore() {
        audioEffectStore = AudioEffectStore.shared()
        deviceStore = DeviceStore.shared()
    }

    override fun addObserver() {
    }

    override fun removeObserver() {
    }

    private fun initView() {
        bindViewId()
        initFinishView()
        initChangeVoiceView()
        initReverbView()
        initEarReturnVolumeView()
        initPeopleVolumeView()
        initEnableEarReturnView()
        initEarReturnVolume()
    }

    private fun bindViewId() {
        textEarReturnVolume = findViewById(R.id.tv_ear_return_volume)
        textPeopleVolume = findViewById(R.id.tv_people_volume)
    }

    private fun initEarReturnVolume() {
        val seekEarReturnVolume = findViewById<SeekBar>(R.id.sb_ear_return_volume)
        seekEarReturnVolume.progress =
            audioEffectStore?.audioEffectState?.earMonitorVolume?.value ?: 0

        seekEarReturnVolume.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                textEarReturnVolume.text = i.toString()
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                audioEffectStore?.setVoiceEarMonitorVolume(seekBar.progress)
            }
        })
    }

    private fun initEnableEarReturnView() {
        val switchEarReturn = findViewById<SwitchCompat>(R.id.sc_enable_ear)
        switchEarReturn.isChecked =
            audioEffectStore?.audioEffectState?.isEarMonitorOpened?.value == true
        switchEarReturn.setOnCheckedChangeListener { button, enable ->
            audioEffectStore?.setVoiceEarMonitorEnable(enable)
        }
    }

    private fun initPeopleVolumeView() {
        textPeopleVolume.text = deviceStore?.deviceState?.captureVolume?.value.toString()

        val seekPeopleVolume = findViewById<SeekBar>(R.id.sb_people_volume)
        seekPeopleVolume.progress = deviceStore?.deviceState?.captureVolume?.value ?: 0
        seekPeopleVolume.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                textPeopleVolume.text = i.toString()
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                deviceStore?.setCaptureVolume(seekBar.progress)
            }
        })
    }

    private fun initEarReturnVolumeView() {
        textEarReturnVolume.text =
            audioEffectStore?.audioEffectState?.earMonitorVolume?.value.toString()
    }

    private fun initFinishView() {
        findViewById<View>(R.id.tv_finish).setOnClickListener { view ->
            onBackButtonClickListener?.onClick()
        }
    }

    private fun initReverbView() {
        val recyclerReverb = findViewById<RecyclerView>(R.id.rv_reverb)
        val adapterReverb = ReverbAdapter(context, audioEffectStore ?: AudioEffectStore.shared())
        val spanCount = adapterReverb.itemCount
        recyclerReverb.layoutManager = GridLayoutManager(context, spanCount)
        recyclerReverb.adapter = adapterReverb
    }

    private fun initChangeVoiceView() {
        val recyclerChangeVoice = findViewById<RecyclerView>(R.id.rv_change_voice)
        val adapterChangeVoice =
            ChangeVoiceAdapter(context, audioEffectStore ?: AudioEffectStore.shared())
        val spanCount = adapterChangeVoice.itemCount
        recyclerChangeVoice.layoutManager = GridLayoutManager(context, spanCount)
        recyclerChangeVoice.adapter = adapterChangeVoice
    }

    private fun reportData(roomId: String) {
        val isVoiceRoom = !TextUtils.isEmpty(roomId) && roomId.startsWith("voice_")
        if (isVoiceRoom) {
            reportEventData(LIVEKIT_METRICS_PANEL_SHOW_VOICE_ROOM_AUDIO_EFFECT)
        } else {
            reportEventData(LIVEKIT_METRICS_PANEL_SHOW_LIVE_ROOM_AUDIO_EFFECT)
        }
    }

    interface OnBackButtonClickListener {
        fun onClick()
    }
}